tree/ntuple/test/rfield_streamer.cxx

```diff
@@ -305,3 +305,53 @@ TEST(RField, StreamerMergeIncremental)
    EXPECT_TRUE(seenStreamerInfos[2]);
    EXPECT_TRUE(seenStreamerInfos[3]);
 }
+
+TEST(RField, StreamerSchemaEvolution)
+{
+   FileRaii fileGuard("test_ntuple_rfield_streamer_schema_evolution.root");
+   {
+      auto model = RNTupleModel::Create();
+      model->AddField(RFieldBase::Create("f", "OldStreamerName<int>").Unwrap());
+      auto writer = RNTupleWriter::Recreate(std::move(model), "ntpl", fileGuard.GetPath());
+      auto ptrF = writer->GetModel().GetDefaultEntry().GetPtr<OldStreamerName<int>>("f");
+      ptrF->fValue = 137;
+      writer->Fill();
+   }
+
+   auto model = RNTupleModel::Create();
+   model->AddField(RFieldBase::Create("f", "NewStreamerName<int>").Unwrap());
+   auto reader = RNTupleReader::Open(std::move(model), "ntpl", fileGuard.GetPath());
+
+   ASSERT_EQ(1U, reader->GetNEntries());
+   auto ptrF = reader->GetModel().GetDefaultEntry().GetPtr<NewStreamerName<int>>("f");
+   reader->LoadEntry(0);
+   EXPECT_EQ(137, ptrF->fValue);
+}
+
+TEST(RField, StreamerClassMismatch)
+{
+   FileRaii fileGuard("test_ntuple_rfield_streamer_class_mismatch.root");
+   {
+      auto model = RNTupleModel::Create();
+      model->AddField(RFieldBase::Create("f", "TemperatureCelsius").Unwrap());
+      auto writer = RNTupleWriter::Recreate(std::move(model), "ntpl", fileGuard.GetPath());
+      auto ptrF = writer->GetModel().GetDefaultEntry().GetPtr<TemperatureCelsius>("f");
+      ptrF->fValue = 100.;
+      writer->Fill();
+   }
+
+   auto model = RNTupleModel::Create();
+   model->AddField(RFieldBase::Create("f", "TemperatureKelvin").Unwrap());
+   auto reader = RNTupleReader::Open(std::move(model), "ntpl", fileGuard.GetPath());
+
+   ASSERT_EQ(1U, reader->GetNEntries());
+   auto ptrF = reader->GetModel().GetDefaultEntry().GetPtr<TemperatureKelvin>("f");
+
+   // TODO(jblomer): this should fail with an exception when we connect the page source
+   ROOT::TestSupport::CheckDiagsRAII diagRAII;
+   diagRAII.requiredDiag(kError, "TBufferFile::ReadVersion", "Could not find the StreamerInfo with a checksum of",
+                         false /* matchFullMessage */);
+   diagRAII.requiredDiag(kError, "TBufferFile::CheckByteCount", "object of class TemperatureKelvin read too few bytes",
+                         false /* matchFullMessage */);
```

Comment:
I don't understand those error message.  What is the case being tested here?

Reply:
I write a `TemperatureCelsius` and then try to read it as `TemperatureKelvin`.

Reply:
Humm ... why isn't the error message closer to "Can not convert (or read) a `TemperatureCelsius` (in)to a `TemperatureKelvin`" ?   Without the corresponding checks  the failure could be 'serious'.  If I understood correctly, the user could have a class `A` with `version 2` which contains a `long` then request the reading a class `B` with also `version 2` which contains a pointer the the I/O would succeed but not apply any check nor conversion and thus have a random number stored in the pointer.  Did I miss something?

Reply:
I think a proper exception is part of the `TODO` written above. I don't understand your example, how is this different from the case tested here that even has compatible class layouts?

Reply:
I guess then we are not using the `TClass` / `TBufferFile` optimally. Currently, when we read the data from a streamer field, we do
```
TBufferFile buffer(TBuffer::kRead, nbytes);
// Fill buffer.Buffer() with nbytes of data
fClass->Streamer(to, buffer);
```

Now, if the `TClass` instance pointed to by `fClass` is not compatible with the streamed data on disk, we get the error messages tested for in the test. How can I make `TClass` report a better error message?

Reply:
> Currently, when we read the data from a streamer field, we do

The check is meant to be done before hand.   RNTuple/TTree know what is stored in the column/branch and what the user is passing and needs to check compatibility of the two.  For TTree, this is done in `Tree::CheckBranchAddressType`, in particular the lines:
```
       ptrClass->GetSchemaRules()->HasRuleWithSourceClass( expectedClass->GetName() ) ) {
```
and (for collection)
```
             inmemValueClass->GetSchemaRules()->HasRuleWithSourceClass(onfileValueClass->GetName() ) )
```

Reply:
Ok. This doesn't seem to be a one-liner and it is a different topic than the one addressed in this PR. This PR is supposed to make the streamer field work if the types _are_ compatible. For better handling of incompatible types, I created https://github.com/root-project/root/issues/18611

Reply:
Fair enough.  This PR is a step forward.   As a side note, in case the class are versions and the on-file class and the unrelated class (in case of user error) have the same class versions, we would not see the message `Could not find the StreamerInfo with a checksum of` but instead the streaming will proceed with the wrong StreamerInfo with a set of possible outcomes: best case scenario we see the `read too few bytes` or `read too many bytes` output, in other cases (if one class layout has pointers in different places (and/or missing/added) than the other class layout) then we could get a crashes and in the 'worst' case scenario the 2 layout have the same number of bytes in their onfile representations and we get a silent failure (wrong data).